gits/terminal.py

```diff
@@ -213,46 +252,66 @@ def _default_rendition(self):
+        self._sgr &= ~(color_bits << 40)  # clear color bits
+        self._sgr |= new_color_bits << 40  # update bg and fg colors
+
+    def _set_fg_color(self, color):
+        """Sets foreground color. """
+        color_bits, _ = divmod(self._sgr, MAGIC_NUMBER)
+        bg, _ = divmod(color_bits, 16)
+
+        # Switch to the bright color scheme if necessary.
+        if not self._normal_mode:
+            color += 8
+
+        new_color_bits = bg * 16 + color
+        self._sgr &= ~(color_bits << 40)  # clear color bits
+        self._sgr |= new_color_bits << 40  # update bg and fg colors
+
     def _cap_set_color_pair(self, p1, p2):
         if p1 == 0 and p2 == 10:  # sgr0
-            self._sgr = MAGIC_NUMBER
+            self._sgr = BLACK_AND_WHITE
         elif p1 == 39 and p2 == 49:  # op
-            self._sgr = MAGIC_NUMBER
+            self._sgr = BLACK_AND_WHITE
         else:
             self._cap_set_color(p1)
             self._cap_set_color(p2)
 
-    def _cap_set_color(self, colour):
-        if colour == 0:
-            self._sgr = MAGIC_NUMBER
-        elif colour == 1:  # bold
-            self._sgr = (self._sgr | 0x08000000)
-        elif colour == 2:  # dim
-            pass
-        elif colour == 4:  # smul
+    def _cap_set_color(self, color):
+        if color == 0:
+            self._sgr = BLACK_AND_WHITE
+        elif color == 1:  # bold
+            self._set_bit(BOLD_BIT)
+            self._cap_set_color(37)  # bold also means extra bright
+        elif color == 2:  # dim
             pass
-        elif colour == 5:  # blink
+        elif color == 4:  # smul, start underscore mode
+            self._set_bit(UNDERLINE_BIT)
+        elif color == 5:  # blink
+            self._set_bit(BLINK_BIT)
+        elif color == 7:  # smso or rev
             pass
-        elif colour == 7:  # smso or rev
-            self._sgr = 0x70000000
-        elif colour == 10:  # rmpch
+        elif color == 10:  # rmpch
             pass
-        elif colour == 11:  # smpch
+        elif color == 11:  # smpch
             pass
-        elif colour == 24:  # rmul
+        elif color == 24:  # rmul, end underscore mode
+            self._claen_bit(UNDERLINE_BIT)
+        elif color == 27:  # rmso
             pass
-        elif colour == 27:  # rmso
-            self._sgr = MAGIC_NUMBER
-        elif 30 <= colour <= 37:  # setaf
-            c = colour - 30
-            self._sgr = (self._sgr & 0xf8ffffff) | (c << 24)
-        elif colour == 39:
-            self._sgr = MAGIC_NUMBER
-        elif 40 <= colour <= 47:  # setab
-            c = colour - 40
-            self._sgr = (self._sgr & 0x0fffffff) | (c << 28)
-        elif colour == 49:
-            self._sgr = MAGIC_NUMBER
+        elif 30 <= color <= 37:  # setaf
+            self._set_fg_color(color - 30)
+        elif color == 39:
+            self._sgr = BLACK_AND_WHITE
+        elif 40 <= color <= 47:  # setab
+            self._set_bg_color(color - 40)
```

Comment:
I'm so happy to see that all this stuff with bit manipulation went to separate method 👍